pkg/networkmanager/wireguard.jsx

```diff
@@ -0,0 +1,257 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2023 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React, { useContext, useEffect, useState } from 'react';
+import cockpit from 'cockpit';
+import { Button } from '@patternfly/react-core/dist/esm/components/Button/index.js';
+import { ClipboardCopy } from '@patternfly/react-core/dist/esm/components/ClipboardCopy/index.js';
+import { EmptyState, EmptyStateBody } from '@patternfly/react-core/dist/esm/components/EmptyState/index.js';
+import { FormGroup, FormFieldGroup, FormFieldGroupHeader } from '@patternfly/react-core/dist/esm/components/Form/index.js';
+import { Grid } from '@patternfly/react-core/dist/esm/layouts/Grid/index.js';
+import { InputGroup } from '@patternfly/react-core/dist/esm/components/InputGroup/index.js';
+import { MinusIcon } from '@patternfly/react-icons';
+import { TextInput } from '@patternfly/react-core/dist/esm/components/TextInput/index.js';
+
+import { Name, NetworkModal, dialogSave } from "./dialogs-common";
+import { ModelContext } from './model-context';
+import { useDialogs } from 'dialogs.jsx';
+
+import './wireguard.scss';
+
+const _ = cockpit.gettext;
+
+function addressesToString(addresses) {
+    return addresses.map(address => address[0] + "/" + address[1]).join(",");
+}
+
+function stringToAddresses(str) {
+    return str.split(",").map(strAddress => {
+        const parts = strAddress.split("/");
+        if (parts.length > 2) {
+            throw new Error(_("Addresses is not formatted correctly"));
+        }
+        const [ip, prefix] = parts;
+        const defaultPrefix = "32";
+        // For WireGuard interfaces gateway should not be set
+        // The gateway may be left as 0 if no gateway exists
+        const gateway = "0.0.0.0";
```

Comment:
Yes the IPv4 settings `addresses` property requires the values (uint32 ip, uint32 prefix, uint32 gateway).

The network page converts ip `"192.168.12.34"` to decimal form `3232238626` then passes that to NM. Same for gateway.

So, `"0.0.0.0"` becomes `0`.

"0.0.0.0" == 0 == No gateway

So NM accepts `(3232238626, 32, 0)` instead of `("192.168.12.34", 32, "0.0.0.0")`

This is confusing I know. That's why I proposed (and gave it a try before) to get rid of using the deprecated `addresses` property.

From the docs:
> [addresses] Deprecated in favor of the 'address-data' and 'gateway' properties, but this can be used for backward-compatibility with older daemons.

We can pass an IP string directly to 'address-data' and 'gateway' properties, without converting them to decimal first unlike in 'addresses'. And let the dbus service throw error if the passed value isn't valid. We can probably get rid of all the code that converts IP string to decimal and vice-versa (for both ipv4 and ipv6) giving `.split()`, `.trim()` on null/undefined errors.

Reply:
That sounds nice indeed! But only worth doing if/once we have a new enough NM on all our OSes. I.e. check RHEL 8 and Ubuntu 22.04 (NM version 1.36.4)